File: tuan08/Chuong5_Full_Code/julia/test_bai_tap.jl
using Test, Random
include("CauTruc.jl")
include("BaiLam.jl")
using .CauTruc, .BaiLam

function make_stack(values)
    s = NganXep()
    for x in values; day!(s, x); end
    return s
end
function queue_wrap()
    q = HangDoiVong(4)
    for x in 1:4; them!(q, x); end
    lay!(q); lay!(q); them!(q, 5); them!(q, 6)
    return q
end

@testset "BT1: đếm giá trị" begin
    @test dem_gia_tri(DSLK(), 2) == 0
    L = DSLK([2, 1, 2, 2]); head = L.dau
    @test dem_gia_tri(L, 2) == 3
    @test gia_tri(L) == [2, 1, 2, 2]
    @test L.dau === head
    @test dem_gia_tri(DSLK([1, 3]), 2) == 0
end
@testset "BT2: xoá mọi giá trị" begin
    L = DSLK([2, 2, 1, 2, 3, 2]); p1, p3 = tim(L, 1), tim(L, 3)
    @test xoa_tat_ca!(L, 2) == 4
    @test gia_tri(L) == [1, 3]
    @test L.n == 2
    @test L.dau === p1
    @test p1.ke === p3
    L = DSLK([4, 4, 4])
    @test xoa_tat_ca!(L, 4) == 3
    @test L.dau === nothing
    @test gia_tri(L) == Int[]
    @test xoa_tat_ca!(DSLK(), 4) == 0
    L = DSLK([1, 2]); @test xoa_tat_ca!(L, 8) == 0
    @test gia_tri(L) == [1, 2]
end
@testset "BT3: lấy nhiều theo LIFO" begin
    s = make_stack([10, 20, 30])
    @test lay_nhieu!(s, 2) == [30, 20]
    @test s.a == [10]
    @test lay_nhieu!(s, 0) == Int[]
    @test s.a == [10]
    @test lay_nhieu!(s, 1) == [10]
    @test rong(s)
    @test lay_nhieu!(s, 0) == Int[]
    for k in (-1, 4, 1.5, true)
        s = make_stack([10, 20, 30]); before = copy(s.a)
        @test_throws ArgumentError lay_nhieu!(s, k)
        @test s.a == before
    end
    @test_throws ArgumentError lay_nhieu!(NganXep(), 1)
end
@testset "BT4: xem k phần tử FIFO" begin
    q = queue_wrap(); before = (copy(q.a), q.dau, q.so, q.n)
    @test xem_k(q, 3) == [3, 4, 5]
    @test (q.a, q.dau, q.so, q.n) == before
    @test xem_k(q, 0) == Int[]
    @test xem_k(q, 4) == [3, 4, 5, 6]
    for k in (-1, 5, 1.5, true)
        q = queue_wrap(); before = (copy(q.a), q.dau, q.so, q.n)
        @test_throws ArgumentError xem_k(q, k)
        @test (q.a, q.dau, q.so, q.n) == before
    end
    @test xem_k(HangDoiVong(4), 0) == Int[]
    @test_throws ArgumentError xem_k(HangDoiVong(4), 1)
end
@testset "200 dãy đối chiếu, seed cố định" begin
    rng = MersenneTwister(20260917)
    for _ in 1:200
        a = rand(rng, -3:3, rand(rng, 0:29)); x = rand(rng, -3:3)
        L = DSLK(a)
        @test dem_gia_tri(L, x) == count(==(x), a)
        @test xoa_tat_ca!(L, x) == count(==(x), a)
        @test gia_tri(L) == filter(!=(x), a)
    end
end

File: tuan08/Chuong5_Full_Code/python/test_bai_tap.py
"""Kiểm thử công khai bốn hàm. Bản sinh viên sẽ báo lỗi cho tới khi cài đặt.

Được dùng list làm kết quả đối chiếu trong kiểm thử; không dùng nó thay nút liên kết.
"""
import random
import unittest
from cau_truc import DSLK, NganXep, HangDoiVong
from bai_lam import dem_gia_tri, xoa_tat_ca, lay_nhieu, xem_k


def stack(values):
    s = NganXep()
    for x in values: s.day(x)
    return s


def queue_wrap():
    q = HangDoiVong(4)
    for x in [1, 2, 3, 4]: q.them(x)
    q.lay(); q.lay(); q.them(5); q.them(6)
    return q  # a = [5,6,3,4]; FIFO = [3,4,5,6]; dau = 2.


class TestBaiTap(unittest.TestCase):
    def test_dem_rong(self):
        self.assertEqual(dem_gia_tri(DSLK(), 2), 0)

    def test_dem_trung_khong_doi(self):
        L = DSLK([2, 1, 2, 2]); head = L.dau
        self.assertEqual(dem_gia_tri(L, 2), 3)
        self.assertEqual(L.gia_tri(), [2, 1, 2, 2]); self.assertIs(L.dau, head)

    def test_dem_khong_co(self):
        self.assertEqual(dem_gia_tri(DSLK([1, 3]), 2), 0)

    def test_xoa_lien_tiep_va_giu_nut_con_lai(self):
        L = DSLK([2, 2, 1, 2, 3, 2]); p1, p3 = L.tim(1), L.tim(3)
        self.assertEqual(xoa_tat_ca(L, 2), 4)
        self.assertEqual(L.gia_tri(), [1, 3]); self.assertEqual(L.n, 2)
        self.assertIs(L.dau, p1); self.assertIs(p1.ke, p3)

    def test_xoa_het(self):
        L = DSLK([4, 4, 4]); self.assertEqual(xoa_tat_ca(L, 4), 3)
        self.assertIsNone(L.dau); self.assertEqual(L.gia_tri(), [])

    def test_xoa_rong(self):
        self.assertEqual(xoa_tat_ca(DSLK(), 4), 0)

    def test_xoa_khong_co(self):
        L = DSLK([1, 2]); self.assertEqual(xoa_tat_ca(L, 8), 0)
        self.assertEqual(L.gia_tri(), [1, 2])

    def test_lay_nhieu_thu_tu(self):
        s = stack([10, 20, 30]); self.assertEqual(lay_nhieu(s, 2), [30, 20])
        self.assertEqual(s.a, [10])

    def test_lay_nhieu_khong(self):
        s = stack([10]); self.assertEqual(lay_nhieu(s, 0), [])
        self.assertEqual(s.a, [10])

    def test_lay_nhieu_het(self):
        s = stack([0, 1]); self.assertEqual(lay_nhieu(s, 2), [1, 0])
        self.assertTrue(s.rong())

    def test_lay_nhieu_k_sai_khong_doi(self):
        for k in [-1, 4, 1.5, True]:
            s = stack([10, 20, 30]); before = s.a.copy()
            with self.assertRaises(ValueError): lay_nhieu(s, k)
            self.assertEqual(s.a, before)

    def test_lay_nhieu_rong(self):
        self.assertEqual(lay_nhieu(NganXep(), 0), [])
        with self.assertRaises(ValueError): lay_nhieu(NganXep(), 1)

    def test_xem_k_quay_vong(self):
        q = queue_wrap(); before = (q.a.copy(), q.dau, q.so, q.n)
        self.assertEqual(xem_k(q, 3), [3, 4, 5])
        self.assertEqual((q.a, q.dau, q.so, q.n), before)

    def test_xem_k_khong_va_het(self):
        q = queue_wrap(); self.assertEqual(xem_k(q, 0), [])
        self.assertEqual(xem_k(q, 4), [3, 4, 5, 6])

    def test_xem_k_sai_khong_doi(self):
        for k in [-1, 5, 1.5, True]:
            q = queue_wrap(); before = (q.a.copy(), q.dau, q.so, q.n)
            with self.assertRaises(ValueError): xem_k(q, k)
            self.assertEqual((q.a, q.dau, q.so, q.n), before)

    def test_xem_k_rong(self):
        self.assertEqual(xem_k(HangDoiVong(4), 0), [])
        with self.assertRaises(ValueError): xem_k(HangDoiVong(4), 1)

    def test_danh_sach_ngau_nhien_co_seed(self):
        rng = random.Random(20260917)
        for _ in range(200):
            values = [rng.randint(-3, 3) for _ in range(rng.randrange(30))]
            x = rng.randint(-3, 3); L = DSLK(values)
            self.assertEqual(dem_gia_tri(L, x), values.count(x))
            self.assertEqual(xoa_tat_ca(L, x), values.count(x))
            self.assertEqual(L.gia_tri(), [v for v in values if v != x])

if __name__ == "__main__": unittest.main(verbosity=2)
